zabbix_smartmontools/__init__.py

```diff
@@ -1,150 +1,204 @@
         else:
-            startStopResult = 'UNKNOWN'
-
-        sender.append('%s smartctl.value[%s,startStop] "%s"' % (host, dR, startStopResult))
-
-        startStopMaxRe = re.search(r'^Recommended maximum start stop count:\s+(\d+)|^Specified cycle count over device lifetime:\s+(\d+)', p, re.M | re.I)
-        if startStopMaxRe:
-            if startStopMaxRe.group(1):
-                startStopMaxResult = startStopMaxRe.group(1)
-            elif startStopMaxRe.group(2):
-                startStopMaxResult = startStopMaxRe.group(2)
+            serial = None
+        if config['mode'] == 'serial' and serial:
+            identifier = serial
+        else:
+            identifier = device
+        jsonData.append({'{#DDRIVESTATUS}': identifier})
+        if not serial in reported_serials or not config['skipDuplicates']:
+            jsonData.append({"{#DISKID}": identifier})
+            jsonData.append({"{#DISKIDSAS}": identifier})
+        if serial:
+            reported_serials.add(serial)
+    return jsonData
+
+
+# Get the serial number of a hard drive, using the fastest method available on
+# each platform.
+def getSerial(device):
+    """ Get the serial number of a hard drive """
+    serial = None
+    if 'freebsd' in sys.platform:
+        DISK_IDENT_SIZE = 256       # from sys/disk.h
+        DIOCGIDENT = 0x41006489     # from sys/disk.h
+        ident = DISK_IDENT_SIZE * r" "
+        f = open("/dev/%s" % device, 'r')
+        try:
+            serial = ioctl(f, DIOCGIDENT, ident).decode().strip("\x00")
+        except OSError as e:
+            if e.errno == errno.EBADF:
+                # EBADF indicates something that's not a disk, such as a SES
+                # enclosure
+                True
             else:
-                startStopMaxResult = 'UNKNOWN'
-
-            sender.append('%s smartctl.value[%s,startStopMax] "%s"' % (host, dR, startStopMaxResult))
-
-    defectsRe = re.search(r'^Elements in grown defect list:\s+(\d+)', p, re.M | re.I)
-    if defectsRe:
-        sender.append('%s smartctl.value[%s,defects] "%s"' % (host, dR, defectsRe.group(1)))
-
-    poweredHoursRe = re.search(r'^number of hours powered up \=\s+(\d+)', p, re.M | re.I)
-    if poweredHoursRe:
-        sender.append('%s smartctl.value[%s,poweredHours] "%s"' % (host, dR, poweredHoursRe.group(1)))
-
-    nonMediumErrorsRe = re.search(r'^Non-medium error count:\s+(\d+)', p, re.M | re.I)
-    if nonMediumErrorsRe:
-        sender.append('%s smartctl.value[%s,nonMediumErrors] "%s"' % (host, dR, nonMediumErrorsRe.group(1)))
-
-    if sender:
-        error = False
+                raise e
+    elif sys.platform.startswith('linux'):
+        # TODO: is there a way to get this without using a subprocess?
+        cp = subprocess.run(["/sbin/udevadm", "info", "--query=all",
+            "--name=/dev/%s" % device], capture_output=True)
+        for line in cp.stdout.decode().splitlines():
+            match = re.match("ID_SERIAL_SHORT=(\S+)", line)
```

Comment:
```suggestion
            match = re.match("(?:[A-Z]: )?ID_SERIAL_SHORT=(\S+)", line)
            if match:
                serial = match.group(1)
```
Have a leading capital letter on Linux.

Reply:
groan.  Would you mind showing me the output of `udevadm info --query=all --name=/dev/sda` on your system?

Reply:
```
# udevadm info --query=all --name=/dev/sda
P: /devices/pci0000:00/0000:00:1c.0/0000:08:00.0/host6/port-6:0/expander-6:0/port-6:0:0/end_device-6:0:0/target6:0:0/6:0:0:0/block/sda
N: sda
S: disk/by-id/ata-WDC_WD5000AAKS-00UU3A0_WD-WCAYUNNNNNNN
S: disk/by-id/wwn-0x50014ee158a2a30c
S: disk/by-path/pci-0000:08:00.0-sas-exp0x5001517508912000-phy0-lun-0
E: DEVLINKS=/dev/disk/by-id/wwn-0x50014ee158a2a30c /dev/disk/by-id/ata-WDC_WD5000AAKS-00UU3A0_WD-WCAYUNNNNNNN /dev/disk/by-path/pci-0000:08:00.0-sas-exp0x5001517508912000-phy0-lun-0
E: DEVNAME=/dev/sda
E: DEVPATH=/devices/pci0000:00/0000:00:1c.0/0000:08:00.0/host6/port-6:0/expander-6:0/port-6:0:0/end_device-6:0:0/target6:0:0/6:0:0:0/block/sda
E: DEVTYPE=disk
E: ID_ATA=1
E: ID_ATA_DOWNLOAD_MICROCODE=1
E: ID_ATA_FEATURE_SET_AAM=1
E: ID_ATA_FEATURE_SET_AAM_CURRENT_VALUE=254
E: ID_ATA_FEATURE_SET_AAM_ENABLED=0
E: ID_ATA_FEATURE_SET_AAM_VENDOR_RECOMMENDED_VALUE=128
E: ID_ATA_FEATURE_SET_HPA=1
E: ID_ATA_FEATURE_SET_HPA_ENABLED=1
E: ID_ATA_FEATURE_SET_PM=1
E: ID_ATA_FEATURE_SET_PM_ENABLED=1
E: ID_ATA_FEATURE_SET_PUIS=1
E: ID_ATA_FEATURE_SET_PUIS_ENABLED=0
E: ID_ATA_FEATURE_SET_SECURITY=1
E: ID_ATA_FEATURE_SET_SECURITY_ENABLED=0
E: ID_ATA_FEATURE_SET_SECURITY_ENHANCED_ERASE_UNIT_MIN=74
E: ID_ATA_FEATURE_SET_SECURITY_ERASE_UNIT_MIN=74
E: ID_ATA_FEATURE_SET_SMART=1
E: ID_ATA_FEATURE_SET_SMART_ENABLED=1
E: ID_ATA_SATA=1
E: ID_ATA_SATA_SIGNAL_RATE_GEN1=1
E: ID_ATA_SATA_SIGNAL_RATE_GEN2=1
E: ID_ATA_WRITE_CACHE=1
E: ID_ATA_WRITE_CACHE_ENABLED=1
E: ID_BUS=ata
E: ID_MODEL=WDC_WD5000AAKS-00UU3A0
E: ID_MODEL_ENC=WDC\x20WD5000AAKS-00UU3A0\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20
E: ID_PART_TABLE_TYPE=gpt
E: ID_PART_TABLE_UUID=09918386-f507-b247-a15d-308b694e9515
E: ID_PATH=pci-0000:08:00.0-sas-exp0x5001517508912000-phy0-lun-0
E: ID_PATH_TAG=pci-0000_08_00_0-sas-exp0x5001517508912000-phy0-lun-0
E: ID_REVISION=01.03B01
E: ID_SERIAL=WDC_WD5000AAKS-00UU3A0_WD-WCAYUNNNNNNN
E: ID_SERIAL_SHORT=WD-WCAYUNNNNNNN
E: ID_TYPE=disk
E: ID_WWN=0x50014ee158a2a30c
E: ID_WWN_WITH_EXTENSION=0x50014ee158a2a30c
E: MAJOR=8
E: MINOR=0
E: SUBSYSTEM=block
E: TAGS=:systemd:
E: USEC_INITIALIZED=258986
```